Changing XCM config for allow teleport from and to AH

metricaez opened the pull request:
Basic changes that need to be implemented for allowing teleporting from and to AH from Trappist. Remove filter, add AH as trusted teleporter and fix Currency adapter.

hbulgarini commented on runtime/trappist/src/xcm_config.rs:
```diff
@@ -331,7 +337,7 @@ impl pallet_xcm::Config for Runtime {
 	type XcmExecuteFilter = Everything;
 	type XcmExecutor = XcmExecutor<XcmConfig>;
-	type XcmTeleportFilter = Nothing;
+	type XcmTeleportFilter = Everything;
```
hmmm i think we should provide `TrustedTeleporters` here instead of everything right?

metricaez replied:
Since this is mostly focused on what could or should try to be teleported out of Trappist I didn't apply any criteria. Should we limit this ?

metricaez replied:
Makes all the sense, will implement 🤝

metricaez replied:
Implemented allowing only Native assets to be teleported out of Trappist.

hbulgarini commented:
@metricaez: is looking very good! however there is a test failing in the asset registry, could you please take a look?

metricaez commented:
Changes implemented:

- Allow teleports from Trappist, implemented a filter that only allows teleporting HOP.
- Allow teleports of HOP from AH to Trappist
- Changed Stout and Trappist Reserves to check that origin is AH but not the asset itself as it generated conflicts with the multilocation of foreign assets. Asset check is done by Asset Registry.
- Added ParachainIDs as valid junction for asset registration on Asset Registry since it's needed for foreign assets.
- Implement HOP Trader on Stout to buy execution with arbitrary value.

@hbulgarini please give it a look and let me know any change you think is needed, otherwise good to merge from my side.

hbulgarini commented on runtime/trappist/src/xcm_config.rs:
```diff
@@ -245,13 +252,24 @@ pub struct ReserveAssetsFrom<T>(PhantomData<T>);
+pub struct OnlyTeleportNative;
+impl Contains<(MultiLocation, Vec<MultiAsset>)> for OnlyTeleportNative {
+	fn contains(t: &(MultiLocation, Vec<MultiAsset>)) -> bool {
+		if t.1.len() != 1 {
```
```suggestion
t.1.iter().any(|asset| {
    log::trace!(target: "xcm::OnlyTeleportNative", "Asset to be teleported: {:?}", asset);
    match asset {
        MultiAsset { id: xcm::latest::AssetId::Concrete(asset_loc), fun: Fungible(_a) } =>
            matches_prefix(&SelfReserve::get(), asset_loc),
        _ => false,
    }
})
```

I'm in personal conflict with this line: Your proposal is much better in terms of performance and cost of execution since it does not iterate. On the other side, i don't like how the `if t.1.len() != 1` reads and also i have a feeling that this will not escalate much if we need to add another scenario to the teleport ( not clear how a new scenario would fit here).

@stiiifff  what do you think?
